picard/util/pipe.py

```diff
@@ -0,0 +1,359 @@
+        _pipe_names = []
+
+        for dir in self.PIPE_DIRS:
+            if dir:
+                _pipe_names.append(os.path.join(os.path.expanduser(dir),
+                                                        f"{app_name}_v{app_version}_pipe_file"))
+
+        if _pipe_names:
+            return _pipe_names
+
+        raise PipeErrorNoDestination
+
+    def _reader(self) -> str:
+        raise NotImplementedError()
+
+    def _sender(self, message) -> bool:
+        raise NotImplementedError()
+
+    def read_from_pipe(self, timeout_secs: Optional[float] = None) -> List[str]:
+        if timeout_secs is None:
+            timeout_secs = self.TIMEOUT_SECS
+
+        reader = self.__thread_pool.submit(self._reader)
+        out = []
+
+        try:
+            res = reader.result(timeout=timeout_secs)
+            if res:
+                for r in res.split(self.MESSAGE_TO_IGNORE):
+                    if r:
+                        out.append(r)
+
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the file-opening loop
+            self.send_to_pipe(self.MESSAGE_TO_IGNORE)
+
+        if out:
+            return out
+
+        return [self.NO_RESPONSE_MESSAGE]
+
+    def send_to_pipe(self, message: str, timeout_secs: Optional[float] = None) -> bool:
+        if timeout_secs is None:
+            timeout_secs = self.TIMEOUT_SECS
+
+        # we're sending only filepaths, so we have to create some kind of separator
+        # to avoid any potential conflicts and mixing the data
+        message += self.MESSAGE_TO_IGNORE
+
+        sender = self.__thread_pool.submit(self._sender, message)
+
+        try:
+            if sender.result(timeout=timeout_secs):
+                return True
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the sender
+            self.read_from_pipe()
+
+        return False
+
+
+class UnixPipe(AbstractPipe):
+
+    PIPE_DIRS = (
+        os.getenv('XDG_RUNTIME_DIR'),
+        "~/.config/MusicBrainz/Picard/pipes/",
+    )
+
+    def __init__(self, app_name: str, app_version: str, args=None, forced_path=None):
+        super().__init__(app_name, app_version, args, forced_path)
+
+        if not self.path:
+            raise PipeErrorNoPermission
+        elif self.is_pipe_owner:
+            self.__create_pipe()
+
+        self._remove_temp_attributes()
+
+    def __create_pipe(self) -> None:
+        # setting false to set make it true only when really created
+        self.is_pipe_owner = False
+        try:
+            try:
+                # just to be sure that there's no broken pipe left
+                os.unlink(self.path)
+            except FileNotFoundError:
+                pass
+            os.makedirs(os.path.dirname(self.path), exist_ok=True)
+            os.mkfifo(self.path)
+            self.is_pipe_owner = True
+            log.debug("Pipe successfully created: %r", self.path)
+        except PermissionError:
```

Comment:
It would be useful to log the actual exception message here, as a warning (or error?).

```python
except PermissionError as exc:
    log.warning("Couldn't create pipe: %r (%s)", self.path, exc)
```